### PR TITLE
Make dsnow and dsnown optional

### PR DESCRIPTION
**Pull request by dabail10**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [X] Short (1 sentence) summary of your PR: 
Reopened this one. This solves issue #492 
- [X] Developer(s): 
dabail10 (D. Bailey)
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash#286630fdecc8efc4dcf4bd7ecf2315101c69c391
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on CICE or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.
This is the icepack part of making both dsnow and dsnown optional. Later I can add some diagnostics for these in CICE.

### CODE REVIEW

**Review by eclare108213 — APPROVED** [review]

**Review by apcraig — CHANGES_REQUESTED** [review]
Looks like you never copy l_dsnow and l_dsnown back into dsnow and dsnown(n) at the end of subroutine icepack_step_therm1.  Or are these intent(in) only?

**Comment by apcraig** [comment]
Before we merge here, maybe we should make sure they work properly with the Icepack driver and/or CICE with whatever changes might be implemented there.

**Comment by dabail10** [comment]
Running a base_suite in CICE to make sure.

**Comment by dabail10** [comment]
Here are the CICE test results. Everything is bfb.

https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#b3d5813f15a5bd23870f73c07016e719d0041bcc

File: columnphysics/icepack_therm_vertical.F90 [hunk 1]
```diff
@@ -2589,7 +2601,8 @@ subroutine icepack_step_therm1(dt,                        &
          meltbn (n) = c0
          congeln(n) = c0
          snoicen(n) = c0
-         dsnown (n) = c0
+         l_dsnown   = c0
+         if (present(dsnown)) l_dsnown = dsnown (n)
```
Comment (apcraig) [inline columnphysics/icepack_therm_vertical.F90:2605]: If the old implementation, "dsnown(n)" was always set to zero before being passed into thermo_vertical.  Now it is being initialized to the incoming value of "dsnown(n)".  Is this correct?  Why doesn't this change answers?

Comment (dabail10) [reply columnphysics/icepack_therm_vertical.F90:2605]: This confused me as well, but dsnow and dsnown are both intent(inout). I guess the initialization to zero is automatically happening in the driver and we were never using it in the driver. I will fix this though.

Comment (dabail10) [reply columnphysics/icepack_therm_vertical.F90:2605]: I am thinking that dsnown should be intent out only.

Comment (dabail10) [reply columnphysics/icepack_therm_vertical.F90:2605]: Also, dsnown (called dsnow) is initialized to zero in thermo_vertical. Again, it should really only be intent(out).

Comment (apcraig) [reply columnphysics/icepack_therm_vertical.F90:2605]: @dabail10, Please make updates ASAP if possible, would like to get this merged and into weekend testing as first steps to release.  Thanks.

**Comment by dabail10** [comment]
Hopefully I can have something later today.

File: columnphysics/icepack_therm_vertical.F90 [hunk 2]
```diff
@@ -2589,7 +2601,8 @@ subroutine icepack_step_therm1(dt,                        &
          meltbn (n) = c0
          congeln(n) = c0
          snoicen(n) = c0
-         dsnown (n) = c0
+         l_dsnown   = c0
+         if (present(dsnown)) dsnown(n) = c0
```
Comment (apcraig) [inline columnphysics/icepack_therm_vertical.F90:2605]: This line is redundant, you should remove it.

**Comment by apcraig** [comment]
I'm going to merge this then remove the redundant line when I update the version number.

**Review by apcraig — APPROVED** [review]

**Pull request merged**